resolver/addressable_resolver.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resolver
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+
+	"knative.dev/eventing/pkg/reconciler/names"
+	"knative.dev/pkg/apis"
+	pkgapisduck "knative.dev/pkg/apis/duck"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	apisv1alpha1 "knative.dev/pkg/apis/v1alpha1"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/tracker"
+
+	"knative.dev/pkg/injection/clients/dynamicclient"
+)
+
+// ForDestination resolves a Destination into a URI.
+// This is a helper component for resources that references other resources' addresses.
+type ForDestination struct {
+	tracker         tracker.Interface
+	informerFactory pkgapisduck.InformerFactory
+}
+
+// NewForDestination creates and initializes a new resolver.ForDestination.
+func NewForDestination(ctx context.Context, callback func(string)) *ForDestination {
```

Comment:
And then one of these for `NewForDestinations` and `NewForReferences`

Reply:
Do we want a separate component for each? We can have one component that understands how to do both. Destination resolution is a superset of reference resolution.